gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+# UDP(DNS) Health Check for Kubernetes Deployment
+
+| Status        |  Proposed        |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [#3](https://github.com/coredns/rfc/pull/3) |
+| **Author(s)** | Jayesh Sharma (@[wjayesh](https://github.com/wjayesh))  |
+| **Sponsor**   |    |
+| **Updated**   | 2020-03-03                                           |
+| **Obsoletes** |  |
+
+## Objective
+
+CoreDNS is the cluster DNS server for Kubernetes and is very critical for the overall health of the Kubernetes cluster. It is important to monitor the health of CoreDNS itself and restarting or repairing any CoreDNS pods that are not behaving correctly.  
+ 
+While CoreDNS exposes a health check itself in the form of Kubernetes’ `livenessProbe`: 
+
+
+- The health check is not UDP (DNS) based. There have been cases where the health port is accessible (for TCP) but CoreDNS itself isn't (UDP). This protocol difference means that CoreDNS is unhealthy from a cluster standpoint, but the control plane can't see this.  
+- The existing health check is also launched locally (the `kubelet` uses the `livenessProbe`) and the situation could be different for pods accessing it remotely. 
+ 
+## Motivation and Scope
+
+
+This project idea aims to get around limitations on Kubernetes’ health check and build an application that: 
+
+
+- Checks CoreDNS health externally through UDP (DNS), from a remote Golang application. 
+- Restart CoreDNS pods by interacting with Kubernetes API through the Golang application, if the response from the cluster and pod IPs is unsatisfactory.  
+
+Thus, making the state of CoreDNS available externally and reliably is important to ensure important services run as they are expected to. A shell script could’ve been used to automate the steps but shell is not good in error handling and may not be reliable depending on which shell (bash/csh/zsh) to use. A Golang program will be much more robust.  
+ 
+### Deliverables  
+
+
+- A standalone Golang application to check CoreDNS health: 
+  - The binary delivered through this project will be a **flexible** application in that it could choose its approach to the problem depending on the environment it is deployed in and the access that has been provided to it.   
+  - The health check program itself doesn’t necessarily reside in the Kubernetes cluster. 
+  - This binary only concerns itself with the health of CoreDNS deployed on Kubernetes. It doesn’t cover other scenarios. 
+
+
+- Proper documentation for the product explaining steps and the concepts used. 
+ 
+ 
+ 
+
+ 
+## Design Proposal 
+
+
+It is necessary to consider all possible cases because the way any team deploys the binary is their **choice** and we should not be forcing a modus operandi to use the product.  
```

Comment:
@WJayesh , this is run on sentence. It is ok in terms of chat/slack, but not for the RFC. Please consider that the presentation matters too. I remember @yongtang gave you that feedback. You placed it here almost _verbatim_.

By the way, you start your section on design by talking about "side" points, e.g. limitations, assumptions, etc.. Instead, you should start it by talking about the design itself. Later, you could get to the limitations and assumptions. That is, make the reader interested in your idea first!